`tests/common/util`: Implement UChild, an abstraction for std::process::Child, to return UChild from run_no_wait instead

Joining7943 opened the pull request:
UChild provides:
* convenience methods to assert the current or final state of the child process
* other convenience methods for usual actions on the child process
* reduce deadlock situations and provide safer piping into the child process stdin

Changes in UCommand:
* We capture the output of the child process stdout and stderr in a temporary file per default instead of using `Stdio::piped` for stdout and stderr.
* Provide a method `stderr_to_stdout` to redirect stderr to stdout

Note, that some tests needed to be adjusted to the new return value of `UCommand::run_no_wait`.

tertsdiepraam commented on tests/by-util/test_tail.rs:
```diff
@@ -2291,7 +2754,9 @@ fn test_follow_name_move2() {
-        args.push("--use-polling");
+        if i == 0 {
+            args.push("--use-polling");
+        }
```
Was this just a mistake left over from before?

Joining7943 replied:
Don't know what to say. Here's the output of clippy
```bash
$ cargo clippy --features tail --no-default-features --test tests -- -D warnings
...

error: it looks like the same item is being pushed into this Vec                                                                                                                   
    --> tests/by-util/test_tail.rs:2294:9                                                
     |                                                                                   
2294 |         args.push("--use-polling");  
     |         ^^^^               
     |                                      
     = note: `-D clippy::same-item-push` implied by `-D warnings`                        
     = help: try using vec!["--use-polling";SIZE] or args.resize(NEW_SIZE, "--use-polling")                                                                                        
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#same_item_push                                                                
                                                                                         
error: it looks like the same item is being pushed into this Vec                                                                                                                   
    --> tests/by-util/test_tail.rs:2454:9                                                
     |                                                                                                                                                                             
2454 |         args.push("--use-polling");                                                                                                                                         
     |         ^^^^                                                                      
     |                              
     = help: try using vec!["--use-polling";SIZE] or args.resize(NEW_SIZE, "--use-polling")                                                                                        
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#same_item_push                                                                
                                            
error: could not compile `coreutils` due to 2 previous errors
```

I haven't found the original suggestions of clippy appropriate.

tertsdiepraam replied:
Hmm maybe I forgot `-D warnings`. Still, it's a change in logic right?

Joining7943 replied:
> Still, it's a change in logic right?

Not here, I guess. The push into the args vec takes place at the end of the iteration, so the second push (which I changed to not happen anymore) is never applied. Even if so, using `--use-polling` twice would produce an usage error. I'm pretty sure this was a mistake when setting up the test case.

Joining7943 replied:
Can't help it but I guess we've got the same version problem here.

Joining7943 replied:
Found a CI run where you can see this failing in the ci:style/lint:ubuntu-latest https://github.com/uutils/coreutils/actions/runs/3526285240/jobs/5914032057

tertsdiepraam replied:
Maybe the fix wasn't good enough. Let's leave it

tertsdiepraam commented on tests/by-util/test_tail.rs:
```diff
@@ -30,6 +37,462 @@ static FOLLOW_NAME_SHORT_EXP: &str = "follow_name_short.expected";
+impl Drop for TestRunner {
+    fn drop(&mut self) {
+        if let Some(mut child) = self.child.take() {
+            let _ = child.kill();
```
Is this not done automatically at `drop`? (Not implying that it is, I genuinely don't know)

tertsdiepraam replied:
Yeah can't hurt I suppose

tertsdiepraam reviewed (commented):
Excellent! Could you add or refactor some tests to show why it's better than the normal test library? If you're adding a new test, please also show an example of what it would need to look like without the `TestRunner`.

Some first observations:
- The commands should probably still be a `UCommand` that you pass to the `TestRunner` so that we don't need to specify each possible option in the `TestRunner` struct, provide builder methods for them, etc.. The way you've written it now feels error prone, because we need to keep the implementation of `TestRunner` in sync with the implementation of `tail` itself.
- I'd like a stronger distinction between a command and a command result, like what `UCommand`/`CmdResult` has. Functions like `assert_stdout_matches` should not be implemented on the runner.
- In general, I would advise you to not make this `TestRunner` do too much. It's doing argument handling, setting up the fixtures, running and managing a process, and checking the output. That's a big burden to put on one struct. It seems to me that your focus should be in managing the child process, because we have acceptable abstractions in place for the other jobs (`UCommand`, `AtPath`, `CmdResult`). To take it to an extreme, maybe it's possible to include most of the functionality as just a wrapper around `std::process::Child`? All the stuff like getting the intermediate `stdout`, `kill_asserted`, `assert_alive` should fit in there. Other things, like `stderr_to_stdout` could actually be interesting for other utils, too, so that could go into `UCommand`.
- It's unclear to me what `run_cmd` is for. Note that you can already call `sh` (or other commands) quite easily with `test_scenario.ccmd("sh")` and you get all the `UCommand` functionality.
- I don't like how it uses things from `tail`. The `FollowMode` and `Signum` could just be supplied as strings, instead. `tail::text` seems to be unused altogether, not just on Windows.

Joining7943 commented on tests/by-util/test_tail.rs:
```diff
@@ -30,6 +37,462 @@ static FOLLOW_NAME_SHORT_EXP: &str = "follow_name_short.expected";
+        assert_eq!(stderr, expected_stderr, "assert stderr");
+    }
+
+    fn cleanup_resources(&mut self, ignore_errors: bool) {
```
Ok, no problem

Joining7943 commented:
Thanks for your comments :)

Perhaps something I'm not sure if it's clear, the TestRunner never intends to or will *replace* UCommand, CmdResult etc. and it is delegating to them where it can. Sure, if you like to you can do all the stuff as usual with the TestRunner, although if you don't need stderr_to_stdout or handling and asserting child processes you're maybe better off with the procedure as usual.

The TestRunner provides stderr_to_stdout (also for the usual run), run_no_wait, run_cmd and the assertions around stderr_to_stdout and child processes, together with some convenience around setting up the arguments for the actual run. I hope this gives you a better idea of what the TestRunner provides and maybe answers some of your points.

If the stderr_to_stdout functionality is useful for others and goes into UCommand I'm happy to be able to remove or at least drastically reduce the handling around it in the `TestRunner` :)

> The way you've written it now feels error prone, because we need to keep the implementation of TestRunner in sync with the implementation of tail itself.

I'm not sure which parts are error prone and need to be in sync with tail (maybe besides the two arguments you mentioned, although I think it's more like a feature than a bug).

> It's unclear to me what run_cmd is for. Note that you can already call sh (or other commands) quite easily with test_scenario.ccmd("sh") and you get all the UCommand functionality.

That's something very useful that can be used in run_cmd.

> I'd like a stronger distinction between a command and a command result, like what UCommand/CmdResult has.

I like that.

> tail::text seems to be unused altogether, not just on Windows.

Yep, that's something I've overseen, text is used in the tests later (which currently aren't there) and then the windows comment fits again ;)

tertsdiepraam commented:
> Perhaps something I'm not sure if it's clear, the TestRunner never intends to or will _replace_ UCommand, CmdResult etc. and it is delegating to them where it can.

Yeah this was clear!

> If the stderr_to_stdout functionality is useful for others and goes into UCommand I'm happy to be able to remove or at least drastically reduce the handling around it in the `TestRunner` :)

Yeah I can think of some cases where this might be useful. We've had trouble testing the interleaving of stdout and stderr for `ls` before. Having this in `UCommand` would be useful!

> I'm not sure which parts are error prone and need to be in sync with tail (maybe besides the two arguments you mentioned, although I think it's more like a feature than a bug).

I was thinking about the arguments, specifically. For example, if I add a new argument and want to add it to the `TestRunner`. I have to add a field, at a setter method with some logic in it and then make sure that it's used in `parse_args`, make sure that it's reset in `reset` and add it to the `Default` impl. That's a lot of things to do if the alternative is just using a slice of strings just like it's used on the command line. I could easily see people forgetting to add it to `reset` for instance leading to subtle bugs _in the tests_. This is why I strongly prefer to remove all argument handling from `TestRunner` and stick `UCommand` for that.

Joining7943 commented:
Here's the basics for stderr_to_stdout in UCommand. I have tested it by trying a little bit around and it worked so far. I'm going to write some proper tests. Maybe you can give me a short feedback upfront?

Joining7943 commented:
I definitely like the abstraction :) Sounds really good. It's a big task, but on the long run I think too the outcome would be worth it.

To be honest I'm not really convinced of the first one. You have this cumbersome setup in the test cases and have to care about the tempfile where you actually only want the output. I'm not a big fan of this extra method, too (maybe there's also a nicer name), but the CapturedOutput can return this tempfile too, if this is something you need. 

Alternatively passing something like CapturedOutput into stderr_to_stderr would be something I could live with.

tertsdiepraam commented:
> It's a big task, but on the long run I think too the outcome would be worth it.

Yeah that's what I'm thinking too. 53 uses is is also not _too_ bad and 34 of those uses are in `test_tail.rs` actually. Most tests don't use `run_no_wait`, but only `run` (or `succeeds`, `fails`).

> Alternatively passing something like CapturedOutput into stderr_to_stderr would be something I could live with.

Something like this? (I'm renaming it to `CombinedOutput`, because that makes more sense with this particular usage)
```rust
let out = CombinedOutput::new();
scene.ucmd().stdout_and_stderr(&out).run();
```

I quite like this!

Joining7943 commented:
I like it, too. I'm looking forward to the discussions. 

So, what's the plan, regarding the refactor and `CombinedOutput`? Do we both, CombinedOutput first and then the refactor... ?

tertsdiepraam commented:
Up to you! If you want to do the `UChild` (or another name) thing that'd be awesome, but if you want to start with just combined output that's cool too.

Joining7943 commented:
The name is perfect. I'll figure something out, then.

Joining7943 commented:
Here's a first implementation of `UChild`. I've also refactored 3 tests in `test_tail` so you have some usage examples and if you like to try around yourself. I hope the handling is intuitive, so far. I haven't started with `stdout_and_stderr(CombinedOutput)`. I'm not exactly sure if we actually need it. Maybe you have some feedback before I start with refactoring all tests? I've also added some comments, documentation for orientation and some TODOs and FIXMEs. The deadlock situation described in the FIXME indeed happened with bigger output, so maybe we should fix this, too.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1189,14 +1239,284 @@ impl UCommand {
+    // TODO: improve this method and recognize if the the command already exited and if so return
+    // the code and success from the command. And maybe panic! if the process is still running and
+    // not run together with stderr_to_stdout, since this method will block indefinitely otherwise.
+    pub fn assert(&mut self) -> CmdResult {
```
I don't like this name, because it's not asserting anything. Do we really need this method? I think with `assert_now` and `wait` we have all we need?

Joining7943 replied:
yeah I think it can be removed.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1189,14 +1239,284 @@ impl UCommand {
+    /// Note, that [`CmdResult::code`] and associated methods will panic. Since the process is still
+    /// running there is no return code to assert. Methods like [`CmdResult::success`] will always
+    /// evaluate to true.
+    pub fn assert_now(
```
Maybe it could be called `intermediate_result`? Ideally this method would be unnecessary if the other methods aren't blocking and return intermediate stdout etc.

Joining7943 replied:
What I really like about this assert prefix is how the tests read: `child.assert_now().stdout_only()...`. Since we're writing within a test framework, I think we can use such naming of methods to keep tests more readable. What about `assert_that(..)`?

tertsdiepraam replied:
I understand the idea, but in my opinion that connection is not obvious enough. I would prefer a more descriptive name.

Joining7943 replied:
I honestly don't like name discussions too much, and read is not describing what you're actually doing.

What's more descriptive than (with regard to remove these expected... arguments)
```rust
child.assert_now_that().stdout_only()
``` 

and you're asserting something about the child? I think, the `that` in the name clearly indicates you're expected to assert something after the call to this method.

tertsdiepraam replied:
> I honestly don't like name discussions too much

Same, let's leave it and come back to this once the rest of the API is more clear and whether we actually need this.

Joining7943 replied:
Ok

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1102,19 +1116,50 @@ impl UCommand {
+    /// child process immediately. Do not use this method directly
+    /// if you want to have stderr redirected to stdout. Use
+    /// [`UCommand::run_no_wait_stderr_to_stdout`] instead.
     pub fn run_no_wait(&mut self) -> Child {
```
Do we still need this one or can it be replace by `run_no_wait_child` and rename that one to `run_no_wait`?

Joining7943 replied:
This method will be the method which returns UChild in the end. To avoid misunderstandings: I'll keep the run methods like they are right now until all tests are adjusted to use `run_no_wait_child` and UChild. I want to be able to run the tests during the rewrite. After the tests are rewritten I'll change the `run_no_wait` method to return `UChild` and rename `run_no_wait_child` to `run_no_wait` in all tests. The unnecessary run methods can then be removed.

tertsdiepraam replied:
That's what I expected, but I just wanted to check. Sounds good!

tertsdiepraam reviewed (commented):
Liking this a lot already. The tests looks so clean! I have some suggestions about the API below.

> I haven't started with stdout_and_stderr(CombinedOutput). I'm not exactly sure if we actually need it.

Yeah I don't think we need `stdout_and_stderr(CombinedOutput)` with this.

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1189,14 +1239,284 @@ impl UCommand {
+    /// Reads and returns all bytes from child's stdout. If `stderr` is not redirected to `stdout`
+    /// with [`UCommand::stderr_to_stdout`] and the child process is still running then this
+    /// method will block indefinitely.
+    pub fn stdout_bytes(&mut self) -> Vec<u8> {
```
would love to do that, but all this blocking of stdout and stderr is inherent to the way we capture output from the child via pipe to the master process. Currently, only the `..exact..` methods can be used to capture the output without blocking as long as the child is still running. As a sidenote: If we would use a file per default in set_stdout and set_stderr like we do with stderr_to_stdout all these blocking methods would not block anymore.

tertsdiepraam replied:
What if a method like `stdout_is` would implicitly do a read up to the length of its input to a buffer and check that?

tertsdiepraam replied:
Wait, actually I don't quite understand. Why can't we use `ChildStdout::read` instead of `ChildStdout::read_exact` to make it non-blocking if we don't want an exact byte count? I suppose there is a possibility to not get the entire thing or something?

Joining7943 replied:
> Would there be any downsides to using a file like you suggest?

I indeed think this is the far better option. I'm not a fan of these exact reads, at all. I've used it for now because this was the established method to read from the pipe, if you look at the methods and functions which are already there. I also noticed that the read_exact method isn't very close to the real world, which we actually test against. 

Let's take for example the test `test_tail::test_follow_single`. If you try it on the command line `target/debug/tail -f file_with_data_but_no_newlines` then there is no output but there definitely should be the output of the file visible. So the test is wrong in passing the assertions. I've fixed this error in #3952. The fix is to flush stdout or a writer with stdout after io::copy etc. Actually, the much bigger problem is, that we don't notice such things in the tests because of this exact reading method in which read is called continuously until the buffer is filled. So, yeah I would prefer using files as default for stderr and stdout instead of Stdio::piped() in run_no_wait.

Joining7943 replied:
... and remove these exact reading methods from UChild

Joining7943 replied:
>  Why can't we use ChildStdout::read instead of ChildStdout::read_to_end to make it non-blocking? I suppose there is a possibility to not get the entire output or something?

Yes, as far as I've understood `read`, `read` doesn't make too much promises about the amount of bytes read from the source. And `read` also needs a fixed size buffer. However, I'll try both, the `read` method and the file method in run_no_wait.

Joining7943 replied:
I gave the change from the default `Stdio::piped()` to file based handling in `run_no_wait` a go. I haven't refactored most of the tests yet, so the ci is failing. I ran all tests on a linux platfrom and there were around 60 failing tests. All of them are using the old `run_no_wait` method without `UChild` and are expected to fail right now. Note, that the usual `run` method still works like before, so we are backwards compatible. I also changed some additional behavior:

* each call to `assert_now` or the `stdout()` and `stderr()` methods are consuming the output, so each subsequent call to any of these methods returns the subsequent output. This matches the old behavior were the output was piped and should be the most intuitive default behavior.
* the new consuming behavior is also the default when using `stderr_to_stdout`, so there's no change in the consuming behavior anymore, using this method or not.

What's also cool about using files is that we have access to the complete output if we wish to do so and nothing is really lost. I just haven't implemented it, yet. What's your opinion?

tertsdiepraam replied:
I like it! Though I think we have to make some simplification steps from here. That is, evaluate the functionality and methods that are actually useful for writing tests.

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1189,14 +1239,284 @@ impl UCommand {
+    /// `ignore_errors` is true unwrap() can be safely used.
+    ///
+    /// # Error If `ignore_errors` is false then the error from [`Child::kill`] is returned.
+    pub fn kill(&mut self, ignore_errors: bool) -> io::Result<()> {
```
I like that

Joining7943 commented:
I've refactored all tests and `util` to use `run_no_wait -> UChild`. There are currently some errors happening during the tests I couldn't solve right away. Here's the background:

I've changed piping input to the child process' stdin happen in a separate thread (in `run()` and `run_piped_stdin()`, what's the majority of use cases, currently), so there should be no deadlocks situations anymore because of piping input into the child process within the actual main process. (`UCommand::pipe_in` together with plain `UCommand::run_no_wait` doesn't work this way, currently. I'm working on it to make this happen, but this is not too important, at the moment)

The changed piping changes the test situation a bit (I think to the better) and there are failures in `test_tac` and `test_tee` related to this change. I'm not familiar with the implementation of them, so I'm not sure, but my suspicions are:
* `test_tee`: The failing tests look like they are supposed to fail because of a broken pipe. I'm not sure if they should use the `UCommand::ignore_stdin_write_error()` method or check for the error to happen.
* `test_tac`: The tests fail randomly on windows when using `UCommand::pipe_in`. I suspect, that the reading from stdin in `tac` doesn't work properly, since `UCommand` doesn't block anymore until all stdin is fed into the child process. In a real world scenario piped input also comes from a separate process, so maybe there's something wrong with `tac`?

Are you familiar with `tee` and `tac` and have some feedback?

tertsdiepraam commented:
I can look more into it later, but have you tried with GNU tee and tac instead and do you get the same errors there?

Joining7943 commented:
> I can look more into it later

This would would be great!

> have you tried with GNU tee and tac instead and do you get the same errors there?

No. How can I run the cargo tests with gnu?

tertsdiepraam commented:
Only manually by changing `scene.ucmd() to `scene.cmd("tac")` I think

Joining7943 commented:
Ah ok... I ran the failing test_tee tests:
```
test_tee::linux_only::test_pipe_error_default
test_tee::linux_only::test_pipe_error_exit
test_tee::linux_only::test_space_error_exit
test_tee::linux_only::test_space_error_exit_nopipe
```
with gnu's tail and they fail, too. 

The tac tests fail only on windows pretty randomly. Can't test with original gnu utils there, I think.

Joining7943 commented:
My current state of investigation is:
Regarding the tee tests, I think that `ignore_stdin_write_error` should be fine. This was the situation before `UChild`. 

Regarding the tac tests, I would like to disable them on windows.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+        }
+    }
+
+    fn new_with(tempdir: &Rc<TempDir>) -> Self {
```
This feels unnecessary, especially if we make this struct private and it's not used here.

Joining7943 replied:
Yeah, this is a leftover. I wanted to place the temporary files from `CapturedOutput` into the temporary directory of the test scenario. It turned out, that some tests (I think some tests in `test_ls` were among them) rely on the content of the folder to be the way they set it up. It also works the way it is, so there's no need for this method anymore.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+    }
+}
+
+// TODO: remove after UChild is ready?
```
Yeah let's get rid of them!

Joining7943 replied:
done

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+        output
+    }
+
+    pub fn write_in<T: Into<Vec<u8>>>(&mut self, data: T) -> io::Result<()> {
```
What would be a scenario where you'd use this over `pipe_in`?

Joining7943 replied:
to fake user input like `y` or something like that.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+    /// the child's stdin cannot be taken out.
+    ///
+    /// [`JoinHandle`]: std::thread::JoinHandle
+    pub fn pipe_in<T: Into<Vec<u8>>>(&mut self, content: T) -> JoinHandle<io::Result<()>> {
```
I feel like I would always want the thread to finish writing, so the behaviour of `pipe_in_and_wait`.

Joining7943 replied:
Please have a look at how it is implemented now. Usually you don't have to care about the join, because it's done automatically in the wait methods.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+        result
+    }
+
+    pub fn pipe_in_and_wait_with_output<T: Into<Vec<u8>>>(mut self, content: T) -> Output {
```
I feel like exposing a `with_output` variant is unnecessary, because we can get the output from the `CmdResult` there.

Joining7943 replied:
was needed in `test_tee::run_tee` that way

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+    ///
+    /// # Panics
+    /// If an error occurred during [`Child::try_wait`]
+    pub fn try_alive(&mut self) -> bool {
+        // TODO: return false on error of try_wait ?
```
Imo, `try` implies a `Result`, so `Result<bool>` in this case. Otherwise another name might be better. For `is_alive` I would indeed expect false on an error.

Joining7943 replied:
I've changed that already locally to `is_alive`, which is indeed way better.

tertsdiepraam reviewed (commented):
Love where this is headed. Most of my comments are focused on removing or hiding parts of the API, I think we should make this as simple as possible for contributors to understand (and the test framework has already a very big API).

I'm also gonna say this now, before we move further. You've got a lot of commits here, but the changes are too big that I don't want to squash them all into one. So it'd be great if you could do some cleanup yourself there.

As for `tac`, maybe `pipe_in_and_wait` is wrong and it's some concurrency bug? Maybe it exists without the thread having written anything somehow.

Joining7943 commented:
Sorry, I already advanced locally but didn't want to push since I've actually only got these questions about tee and tac and didnt' want to disturb your investigation on this topic. I think a lot of your comments are outdated. However, we're pretty close to a final review, I guess. Maybe we can wait until I'm ready.

I'm close to 100% sure, we don't have a concurrency bug. Also, why should a concurrecy bug in `pipe_in` only happen with tac on windows? `pipe_in` is used quiet often in all test no matter the platform. Isn't tac actually still under development? The readme states tac isn't finished, yet. A quick glance into tac also shows some unsafe code when reading from stdin. I'm not sure if reading from stdin is done correctly, there.

tertsdiepraam commented:
Alright ping me when you think it's ready!

I've not looked into `tac` itself. A concurrency bug actually would explain why it only falls on Windows, because threads work differently on different platforms, but it was just a wild guess.

Joining7943 commented:
> Alright ping me when you think it's ready!

Ok, thanks. This doesn't mean that I don't like your comments. I just think this is less work for both of us doing it that way, and I'm close to an end anyways.

> I've not looked into tac itself.
Ok, I'm disabling the tac tests on windows using pipe_in and leave a note mentioning this pr.

Maybe it's also worth to note, that we're actually solving a concurrency bug by doing the piping in a separate thread.

Joining7943 commented on tests/common/util.rs:
```diff
@@ -942,6 +944,9 @@ pub struct UCommand {
     limits: Vec<(rlimit::Resource, u64, u64)>,
+    stderr_to_stdout: bool,
+    captured_stdout: Option<CapturedOutput>,
+    captured_stderr: Option<CapturedOutput>,
```
`captured_stdout` and `captured_stderr`, yes. `stderr_to_stdout` is needed for the builder so we can do

```rust
scene.ucmd().stderr_to_stdout().run()
```

tertsdiepraam replied:
Makes sense!

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+    }
+
+    pub fn close_stderr(&mut self) -> &mut Self {
+        if let Some(stderr) = self.raw.stderr.take() {
+            drop(stderr);
+        }
```
Right! I somehow liked the explicitness. It's changed, though.

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+            .expect("Could not pipe into child process. Was it set to Stdio::null()?");
+        thread::spawn(move || {
+            let mut writer = BufWriter::new(stdin);
+            if ignore_stdin_write_error {
+                let _ = writer.write_all(&content);
+                let _ = writer.flush();
+                Ok(())
+            } else {
+                writer
+                    .write_all(&content)
+                    .and_then(|_| writer.flush())
+                    .map_err(|error| {
+                        io::Error::new(
+                            io::ErrorKind::Other,
+                            format!("failed to write to stdin of child: {}", error),
+                        )
+                    })
```
done with a little correction: `if !ignore_stdin_write_error ...`

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+        let output = self.wait_with_output().unwrap();
+        thread.join().unwrap().unwrap();
+        output
+    }
```
the `test_tee::run` method uses it. These tests could be refactored to use pipe_in_and_wait -> CmdResult if liked to. I didn't want to do more than necessary in the tests during this pr.

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+        UChildAssertion::new(self)
+    }
+
+    pub fn make_assertion_with_delay(&mut self, millis: u64) -> UChildAssertion {
+        sleep(Duration::from_millis(millis));
+        self.make_assertion()
+    }
```
convenience and few key strokes less. I simplified the method a bit to the line you mentioned.

tertsdiepraam replied:
Not sure about saving keystrokes :)
```
child.delay(millis).make_assertion()
child.make_assertion_with_delay(millis)
```

More seriously though, I think we should keep simplifying this as much as possible, so I think this can be removed.

Joining7943 replied:
Autocompletion is your friend here ;) This is a very common usage pattern, so it makes perfect sense having the separate functions both together in a convenience function. I mean you can do as you like, although I prefer the convenience function.

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+        })
+    }
+
+    pub fn wait_with_output(mut self) -> io::Result<Output> {
```
No? `child.wait().unwrap()` returns a `CmdResult`. I would prefer keeping it public and make it available, doesn't hurt imho.

tertsdiepraam replied:
Yeah okay maybe not exactly `child.wait().output`, but the `CmdResult` should give you all you should need. I like your deprecation solution though we will need to annotate the tests with allowing the deprecated calls in the meantime.

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+    }
+
+    /// Returns the captured output as [`String`]. See also [`CapturedOutput::output_as_bytes`].
+    pub fn output(&mut self) -> String {
```
yep, can be private

tertsdiepraam replied:
Cool, could you mark these methods private too then?

Joining7943 replied:
sure

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1102,19 +1118,53 @@ impl UCommand {
         self
     }
 
+    // TODO: Add convenience method run_no_wait_with_delay which accept a parameter `delay` which
+    // returns delayed from run_no_wait. A lot of use cases are with some kind of post delay.
+    // Without any delay, the output may be empty because we return immediately.
```
yep

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+        panic!("Killing the child process within 60 seconds failed.")
+    }
+
+    // TODO: Accept AssertionMode as parameter
```
It's not that easy, although I think we match the common use cases the way it is right now.

```rust
ts.ucmd().run() -> CmdResult // where all output is in CmdResult

let child = ts.ucmd().run_no_wait();
// do something and/or child outputs something
child.make_assertion().with_current_output() // assert on the output from beginning to now
// do something and child outputs something
// if child is supposed to exit we can wait
child.wait() // (*) -> CmdResult and do assertions on the output from last call to make_assertion until the exit
// if child doesn't exit but we are finished with it
child.kill() // -> UChildAssertion with which we can use all, current or exact output
```

I'm not sure if it's ever used so it was still a TODO, but I wanted to improve on the `(*)` case, so you're able to use all output from start to end of the child process or the exact output and not only the current.

tertsdiepraam replied:
Oh I see. I think that's a design question we need to answer once we find a need for it. In general, with `wait()` and `kill()` I'm expecting to get everything I can get and it would be an extremely rare situation where I might need to do something more complex. In which case I case also use the raw `Child` still.

Joining7943 replied:
I'll try something, here.

Joining7943 replied:
I think I found a simple solution for the `(*)` case. It would introduce 2 public methods in `UChild`. Interested or keep it the way it is for now?

Joining7943 replied:
In an afterthought, I don't think, there's much gain from a usage perspective. Closing this?

tertsdiepraam replied:
Yeah resolving this is fine

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+    ignore_stdin_write_error: bool,
+}
+
+// TODO: rename `pipe_in` to `pipe_in_bytes(Vec<u8>)` and use convenience function `pipe_in(&str)`
```
solved it differently by accepting `Into<Vec<u8>>` for `pipe_in`

tertsdiepraam replied:
Oh interesting. I like that!

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1191,12 +1243,525 @@ impl UCommand {
+}
+
+// TODO: rename `pipe_in` to `pipe_in_bytes(Vec<u8>)` and use convenience function `pipe_in(&str)`
+// TODO: Add method `pipe_in_fixture(&str)`: like `pipe_in` but with a fixture
```
ok

Joining7943 commented:
@tertsdiepraam ping :)

Joining7943 commented:
I may have a solution for our problem with the `wait_with_output` and `pipe_in_and_wait_with_output` methods in `UChild`. I think too, they are actually unnecessary and `CmdResult` is the far better choice to work with. 

I'm not sure how the ci and the lints react on it, but if we would mark them with `#[deprecated = Please use ... instead.]`, we can remove them as soon all tests are refactored to use the `wait` and `pipe_in_and_wait` methods which return a `CmdResult` instead of `Output`. What do you think?

We would gain a little bit more time to get the tests adjusted to the new situation. Especially the `tee` tests use `Output` a lot. The `run` method in `test_yes` uses `Output::status.signal()` on unix systems, which isn't available in `CmdResult`. I guess the best choice is to adjust `CmdResult` to store the `std::process::ExitStatus` (or even better `Option<ExitStatus>` because we also use the `CmdResult` while the process is still running) and replace `CmdResult::code` and `CmdResult::success`.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,26 +1221,613 @@ impl UCommand {
+    }
+
+    /// Returns true if the child process is still alive and false otherwise.
+    #[allow(clippy::wrong_self_convention)]
```
I think these annotations can be removed, because the self convention is correct with the name change.

Joining7943 replied:
I think you're mixing something up here. The wrong self convention comes with naming the method `is_...`.

tertsdiepraam replied:
I can't reproduce it locally, so I'm confused. I'm running
```
cargo clippy --all-targets -- -D warnings
```

As far as I can tell from the source code of the clippy lint, it should allow `is_alive(&mut self)`, but only warns for `is_alive(self)`. So we should be good right? Or am I missing something?

Joining7943 replied:
dunno. Here are the docs for this warning https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention. I'm getting this warning locally for `is_alive(&mut self)`

Joining7943 replied:
I'm using `cargo clippy --features unix --test tests`

tertsdiepraam replied:
The docs are why I'm confused. Here's a playground link that shows that mutable reference should be fine: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d0107deeee912829ef29bf6ef5385daa

In the last year, it appears this lint has been changed twice:
- Disallowing `&mut self`: https://github.com/rust-lang/rust-clippy/issues/8142, https://github.com/rust-lang/rust-clippy/pull/8208
- And then allowing it again: https://github.com/rust-lang/rust-clippy/issues/8513, https://github.com/rust-lang/rust-clippy/pull/8738

So maybe your clippy is out of date?

Joining7943 replied:
I'm using the clippy version from the current min rust version 1.60.0 which is up-to-date  although it's not the latest version around

```
$ rustup show active-toolchain
1.60.0-x86_64-unknown-linux-gnu

$ cargo clippy --version
clippy 0.1.60 (7737e0b 2022-04-04)
```

tertsdiepraam replied:
Oh that explains it I guess. The CI uses the latest stable version and we have a min rust version setting for it to ensure compatibility with msrv, so let's follow the CI.

Joining7943 replied:
ok. The `#[allow...wrong_self_convention]` attributes are removed from the is_* methods.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,26 +1221,613 @@ impl UCommand {
+        }
+    }
+
+    /// Reads and returns all output from start of the child process until now. Each subsequent call
+    /// of any of the methods below will operate on the subsequent output of the child process. Note
+    /// that this function will work like [`UChild::stdout_bytes`] if the output wasn't captured
+    /// with [`CapturedOutput`], for example if [`UCommand::set_stdout`] was used with
+    /// [`Stdio::piped`].
```
Two things about this behaviour.
1. If it would be the same, I would assume it's user error and we should maybe panic if this is called without a captured stdout, but I'm not sure.
2. If you do keep it, the behaviour should be documented on `stdout_all` and `stderr_all` as well.

Joining7943 replied:
Let's panic then. It's definitely an usage error.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,26 +1221,613 @@ impl UCommand {
+    /// * [`UChild::pipe_in_and_wait`]
+    /// * [`UChild::pipe_in_and_wait_with_output`]
+    ///
+    /// Usually, there's no need to join manually but if needed, the [`UChild::join`] method can be
```
You've called it `join_in`, though I indeed like just `join` better.

Joining7943 replied:
oh, I see. Then let's call it `join`.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,26 +1221,613 @@ impl UCommand {
+    }
+
+    /// Convenience method for [`UChild::pipe_in`] and then [`UChild::wait`]
+    pub fn pipe_in_and_wait<T: Into<Vec<u8>>>(mut self, content: T) -> CmdResult {
```
Same as `make_assertion_with_delay`. I guess it's personal preference, but I'd like to keep the API small, so I think these are not necessary. I do like the `try_` methods though, because they fit with the rest of the ecosystem.

Joining7943 replied:
I really like having convenience methods for common usage patterns. This one saves you writing two lines instead of just one. The `wait()` methods consumes `UChild`, so it's not possible to write `self.pipe_in("something").wait().unwrap()` without getting a compiler error.

tertsdiepraam replied:
Oh that's annoying. I agree then, though it is something I'd like to clean up later by making `pipe_in` take ownership of `self`, but that would be a big breaking change.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,26 +1221,613 @@ impl UCommand {
+
+    /// Write some bytes to the child process stdin. This function is meant for small data and
+    /// faking user input like typing a `yes` or `no`. This function blocks until all data is
+    /// written but can be used multiple times in contrast to [`UChild::pipe_in`].
```
Excellent explanation, thanks!

A quick rustdoc tip, the first line is usually for a small summary, followed by a blank line a longer explanation. It will use the short summary in the pages for modules for example. I suppose though that we don't really need to worry about that in this case though because I don't even know how to run rustdoc for this file.

Joining7943 replied:
> A quick rustdoc tip, the first line is usually for a small summary, followed by a blank line a longer explanation.

Thanks. It's a good idea to separate the first sentence from the rest, if the docs can be built or not. I had the test writers in mind and the docs popup in the IDE.

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,26 +1221,613 @@ impl UCommand {
+    }
+
+    /// Closes the child process stdout. Note this will have no effect if the output was captured
+    /// with [`CapturedOutput`] what is the default if [`UCommand::set_stdout`] wasn't called.
```
Nit: "_which_ is the default`

Joining7943 replied:
fixed

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -1191,26 +1221,613 @@ impl UCommand {
+    Current,
+    Exact(usize, usize),
+}
+pub struct UChildAssertion<'a> {
```
At this point, I'm wondering if we still need this? It's always `.make_assertion()` followed by `with_all_output`, `with_exact_output` or `with_current_output`, so those could be methods directly on `UChild`, right? `UChildAssertion` could still be used internally if that makes things easier, but I think we can hide it.

There's also `make_assertion().is_alive`, which could maybe just be `assert_alive` on `UChild`?

tertsdiepraam replied:
Oh I see we need something for `kill` to return too. This should be just all the output in my opinion, so the same as `child.make_assertion().with_all_output()`

Joining7943 replied:
I disagree, here. I really don't want having the assertions within `UChild`. The assertions are separated from actions, are tidied up and just play nicely together in `UChildAssertion`. This makes writing and reading the tests more convenient and fun. It's also a good place for adding new assertions if needed.

> This should be just all the output in my opinion, so the same as child.make_assertion().with_all_output()

No, not if you've asserted `with_current_output` or similar before you've called `wait` or `kill`. You'll just have to assert the rest of the output since the last assertion and not all output again. Please believe me. This assertion pattern is very common in `tail` and the way it is right now is nice and intuitive to work with. Note, if you haven't used any assertions since the start of the process then you'll get all output after a wait!

I would prefer leaving it to the user of `UChild` what to do after calling `kill()`. 

Overall, we should not try to be too closed on the usage of the api. We can't see in advance all needs of the tests and of `UChild`. The usage of the api should be up to the user of it and how he wants/needs to write the tests instead of enforcing a specific way of usage.

tertsdiepraam replied:
Okay, I understand the difference, but I think that difference might not be super clear in the API at the moment, but I don't have any ideas to improve it. Let's leave it like you've written it now.

tertsdiepraam reviewed (commented):
Wow, these changes you've made are indeed great. I've added a few more comments. Please also check some of the unresolved previous comments still.

I'm concerned about `tac` though. I thought it was just about a few tests, but this is a lot to disable. I would be a bit less concerned about it if you could add some adds of the form in which is it used in `tac` and show that they succeed. So this pattern:
```rust
new_ucmd!()
    .pipe_in(...)
    .run()
    .stdout_is(...);
```
Could be with `cat` for example. Just to be absolutely sure it's `tac` and not the framework.

Joining7943 commented:
> Wow, these changes you've made are indeed great.

Thanks. I was trying to make the api rich and convenient to use. I think the tests can also be written much more readable, now.

> Please also check some of the unresolved previous comments still.

Some of you're previous comments are outdated, maybe you haven't noticed? I'll answer them with something short, so you'll see.

> Could be with cat for example. Just to be absolutely sure it's tac and not the framework.

Honestly, I wouldn't be concerned about `tac`. I've already added a test with `pipe_in` in `tests/common/util`. However, I'm adding an additional one. Please, take also into consideration, that we're using `pipe_in` and `pipe_in_fixture` in 624 

```bash
grep -r pipe_in tests/by-util | wc -l
```

tests, but only the tac tests fail on windows intermittently.

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1166,7 +1217,8 @@ impl UCommand {
     pub fn run_piped_stdin<T: Into<Vec<u8>>>(&mut self, input: T) -> CmdResult {
-        self.pipe_in(input).run()
+        self.bytes_into_stdin = None;
+        self.run_no_wait().pipe_in_and_wait(input)
```
outdated?

Joining7943 commented on tests/common/util.rs:
```diff
@@ -1102,19 +1118,53 @@ impl UCommand {
     /// Spawns the command, feeds the stdin if any, and returns the
-    /// child process immediately.
-    pub fn run_no_wait(&mut self) -> Child {
+    /// child process immediately. Do not use this method directly
+    /// if you want to have stderr redirected to stdout. Use
+    /// [`UCommand::run_no_wait_stderr_to_stdout`] instead.
```
outdated?

tertsdiepraam commented on tests/by-util/test_cp.rs:
```diff
@@ -2330,10 +2330,7 @@ fn test_copy_contents_fifo() {
-    assert!(output.status.success());
-    assert!(output.stdout.is_empty());
-    assert!(output.stderr.is_empty());
+    child.wait().unwrap().no_stdout().no_stderr().success();
```
Nice!

tertsdiepraam commented on tests/by-util/test_sort.rs:
```diff
@@ -978,7 +978,7 @@ fn test_sigpipe_panic() {
     child.close_stdout();
     assert_eq!(
-        String::from_utf8(child.wait_with_output().unwrap().stderr),
+        String::from_utf8(Vec::from(child.wait().unwrap().stderr())),
```
Isn't this just `stdout_empty()`?

Joining7943 replied:
Nice! The whole assert_eq! can be exchanged with `child.wait().unwrap().no_stderr()`

tertsdiepraam reviewed (commented):
I agree with you on these points you've made. Excellent work! We just need a few `#[allow(deprecated)]` on the uses of `wait_with_output`.

tertsdiepraam commented:
> Honestly, I wouldn't be concerned about tac.

Thanks for adding the test, however silly it might be in practice. I've just looked at `tac` and it's using mmap, which specifies in its docs:

> All file-backed memory map constructors are marked `unsafe` because of the potential for *Undefined Behavior* (UB) using the map if the underlying file is subsequently modified, in or out of process. Applications must consider the risk and take appropriate precautions when using file-backed maps. Solutions such as file permissions, locks or process-private (e.g. unlinked) files exist but are platform specific and limited.

So the failure might be caused by the change in piping in the data, but the problem is then that we're not careful enough in `tac`.

Joining7943 commented:
Ok. Cool.

> We just need a few #[allow(deprecated)] on the uses of wait_with_output.

I've just tried something and ran the ci/clippy job with `clippy ... -- -Dwarning -Adeprecated`, so the ci doesn't fail anymore because of deprecation warnings. But, we would then have the deprecation warnings locally. What do you think?

tertsdiepraam commented:
I prefer just having the annotations in the code. I don't want the warnings for every contributor locally either and I'd like to see other deprecation warnings

Joining7943 commented:
Ok no problem.

Joining7943 commented:
> We just need a few #[allow(deprecated)] on the uses of wait_with_output.

Done.

tertsdiepraam commented:
> Are we ready, here?

Almost, it all looks great and it's just the clippy things now!

Joining7943 commented on tests/by-util/test_tac.rs:
```diff
@@ -7,6 +7,8 @@ fn test_invalid_arg() {
 }
 
 #[test]
+// FIXME: See https://github.com/uutils/coreutils/pull/4136
```
Here's the issue #4204

tertsdiepraam replied:
Thanks!

tertsdiepraam commented on tests/common/util.rs:
```diff
@@ -361,6 +362,7 @@ impl CmdResult {
         self.no_stderr().stdout_is_bytes(msg)
     }
 
+    // TODO: implement same functionality asserting as String instead
```
Okay, truly the last thing I can find and then it's ready to merge in my opinion if the CI passes. Do we still need this? It makes sense to me that the assertion with a fixture always compares bytes.

Joining7943 replied:
I removed the TODO. However, I've got some ideas to improve `CmdResult` a little bit, but `CmdResult` itself shouldn't be part of this pr.

tertsdiepraam reviewed (approved):
Very nice! I'm super happy with how this turned out! Thanks!

Joining7943 commented:
Sure. I'm glad to hear.

sylvestre commented:
looks great indeed, thanks :)
